File: src/main/java/no/nav/sosialhjelp/soknad/consumer/kodeverk/KodeverkService.java
//package no.nav.sosialhjelp.soknad.consumer.kodeverk;
//
//import no.nav.sosialhjelp.soknad.consumer.kodeverk.dto.KodeverkDto;
//import no.nav.sosialhjelp.soknad.consumer.redis.RedisService;
//import org.slf4j.Logger;
//import org.springframework.stereotype.Component;
//
//import java.time.LocalDateTime;
//import java.util.Map;
//
//import static java.time.format.DateTimeFormatter.ISO_LOCAL_DATE_TIME;
//import static no.nav.sosialhjelp.soknad.consumer.kodeverk.dto.BetydningDto.SPRAAKKODE_NB;
//import static no.nav.sosialhjelp.soknad.consumer.redis.CacheConstants.KODEVERK_LAST_POLL_TIME_KEY;
//import static no.nav.sosialhjelp.soknad.consumer.redis.CacheConstants.KOMMUNER_CACHE_KEY;
//import static no.nav.sosialhjelp.soknad.consumer.redis.CacheConstants.LANDKODER_CACHE_KEY;
//import static no.nav.sosialhjelp.soknad.consumer.redis.CacheConstants.POSTNUMMER_CACHE_KEY;
//import static org.slf4j.LoggerFactory.getLogger;
//
//@Component
//public class KodeverkService {
//
//    private static final Logger logger = getLogger(KodeverkService.class);
//    private static final long MINUTES_TO_PASS_BETWEEN_POLL = 60;
//
//    private final KodeverkConsumer kodeverkConsumer;
//    private final RedisService redisService;
//
//    public KodeverkService(KodeverkConsumer kodeverkConsumer, RedisService redisService) {
//        this.kodeverkConsumer = kodeverkConsumer;
//        this.redisService = redisService;
//    }
//
//    public String getKommunenavn(String kommunenummer) {
//        var kommuneKodeverk = hentKodeverkFraCacheEllerConsumer(KOMMUNER_CACHE_KEY);
//        return finnFoersteTermForKodeverdi(kommuneKodeverk, kommunenummer);
//    }
//
//    public String gjettKommunenummer(String kommunenavn) {
//        var kommuneKodeverk = hentKodeverkFraCacheEllerConsumer(KOMMUNER_CACHE_KEY);
//        return finnKodeverdiForFoersteTerm(kommuneKodeverk, kommunenavn);
//    }
//
//    public String getPoststed(String postnummer) {
//        var postnummerKodeverk = hentKodeverkFraCacheEllerConsumer(POSTNUMMER_CACHE_KEY);
//        return finnFoersteTermForKodeverdi(postnummerKodeverk, postnummer);
//    }
//
//    public String getLand(String landkode) {
//        var landkoderKodeverk = hentKodeverkFraCacheEllerConsumer(LANDKODER_CACHE_KEY);
//        var land = finnFoersteTermForKodeverdi(landkoderKodeverk, landkode);
//        return formaterLand(land);
//    }
//
//    private KodeverkDto hentKodeverkFraCacheEllerConsumer(String key) {
//        if (skalBrukeCache()) {
//            var kodeverk = hentFraCache(key);
//            if (kodeverk != null) {
//                return kodeverk;
//            }
//        }
//        switch (key) {
//            case KOMMUNER_CACHE_KEY:
//                return kodeverkFraConsumerEllerForsoekMedCacheSomFallback(key, kodeverkConsumer.hentKommuner());
//            case POSTNUMMER_CACHE_KEY:
//                return kodeverkFraConsumerEllerForsoekMedCacheSomFallback(key, kodeverkConsumer.hentPostnummer());
//            case LANDKODER_CACHE_KEY:
//                return kodeverkFraConsumerEllerForsoekMedCacheSomFallback(key, kodeverkConsumer.hentLandkoder());
//        }
//
//        return null;
//    }
//
//    private boolean skalBrukeCache() {
//        String timeString = redisService.getString(KODEVERK_LAST_POLL_TIME_KEY);
//        if (timeString == null) {
//            return false;
//        }
//        LocalDateTime lastPollTime = LocalDateTime.parse(timeString, ISO_LOCAL_DATE_TIME);
//        return lastPollTime.plusMinutes(MINUTES_TO_PASS_BETWEEN_POLL).isAfter(LocalDateTime.now());
//    }
//
//    private KodeverkDto hentFraCache(String key) {
//        return (KodeverkDto) redisService.get(key, KodeverkDto.class);
//    }
//
//    private KodeverkDto kodeverkFraConsumerEllerForsoekMedCacheSomFallback(String key, KodeverkDto kodeverk) {
//        if (kodeverk != null) {
//            return kodeverk;
//        }
//        var cachedKodeverk = hentFraCache(key);
//        if (cachedKodeverk != null) {
//            logger.info("Kodeverk-client feilet, men bruker cached kodeverk.");
//            return cachedKodeverk;
//        }
//        logger.warn("Kodeverk feiler og cache [{}] er tom", key);
//        return null;
//    }
//
//    private String finnKodeverdiForFoersteTerm(KodeverkDto kodeverk, String term) {
//        if (kodeverk == null) {
//            return null;
//        }
//
//        return kodeverk.getBetydninger().entrySet().stream()
//                .filter(entry -> {
//                    var betydningDtos = entry.getValue();
//                    if (!betydningDtos.isEmpty() && betydningDtos.get(0) != null && betydningDtos.get(0).getBeskrivelser().get(SPRAAKKODE_NB).getTerm() != null) {
//                        return term.equalsIgnoreCase(betydningDtos.get(0).getBeskrivelser().get(SPRAAKKODE_NB).getTerm());
//                    }
//                    return false;
//                })
//                .findFirst()
//                .map(Map.Entry::getKey)
//                .orElse(null);
//    }
//
//    private String finnFoersteTermForKodeverdi(KodeverkDto kodeverk, String kodeverdi) {
//        if (kodeverk != null && kodeverk.getBetydninger().containsKey(kodeverdi)) {
//            return kodeverk.getBetydninger().get(kodeverdi).get(0).getBeskrivelser().get(SPRAAKKODE_NB).getTerm();
//        }
//        return null;
//    }
//
//    private String formaterLand(String land) {
//        if (land == null) {
//            return land;
//        }
//        String formaterMedSpace = setUpperCaseBeforeRegex(land.toLowerCase(), " ");
//        String formaterMedDash = setUpperCaseBeforeRegex(formaterMedSpace, "-");
//        return setUpperCaseBeforeRegex(formaterMedDash, "/");
//    }
//
//    private String setUpperCaseBeforeRegex(String s, String regex) {
//        String[] split = s.split(regex);
//        StringBuilder sb = new StringBuilder();
//        for (int i = 0; i < split.length; i++) {
//            if (i > 0) {
//                sb.append(regex);
//            }
//
//            if (split[i].equals("og")) {
//                sb.append(split[i]);
//            } else {
//                sb.append(split[i].substring(0, 1).toUpperCase());
//                sb.append(split[i].substring(1));
//            }
//        }
//        return sb.toString();
//    }
//}

File: src/main/kotlin/no/nav/sosialhjelp/soknad/client/kodeverk/KodeverkService.kt
package no.nav.sosialhjelp.soknad.client.kodeverk

import no.nav.sosialhjelp.soknad.client.kodeverk.dto.KodeverkDto
import no.nav.sosialhjelp.soknad.client.redis.KODEVERK_LAST_POLL_TIME_KEY
import no.nav.sosialhjelp.soknad.client.redis.KOMMUNER_CACHE_KEY
import no.nav.sosialhjelp.soknad.client.redis.LANDKODER_CACHE_KEY
import no.nav.sosialhjelp.soknad.client.redis.POSTNUMMER_CACHE_KEY
import no.nav.sosialhjelp.soknad.client.redis.RedisService
import org.slf4j.LoggerFactory
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.Locale

open class KodeverkService(
    private val kodeverkClient: KodeverkClient,
    private val redisService: RedisService
) {

    open fun getKommunenavn(kommunenummer: String): String? {
        val kommuneKodeverk = hentKodeverkFraCacheEllerConsumer(KOMMUNER_CACHE_KEY)
        return finnFoersteTermForKodeverdi(kommuneKodeverk, kommunenummer)
    }

    open fun gjettKommunenummer(kommunenavn: String): String? {
        val kommuneKodeverk = hentKodeverkFraCacheEllerConsumer(KOMMUNER_CACHE_KEY)
        return finnKodeverdiForFoersteTerm(kommuneKodeverk, kommunenavn)
    }

    open fun getPoststed(postnummer: String): String? {
        val postnummerKodeverk = hentKodeverkFraCacheEllerConsumer(POSTNUMMER_CACHE_KEY)
        return finnFoersteTermForKodeverdi(postnummerKodeverk, postnummer)
    }

    open fun getLand(landkode: String): String? {
        val landkoderKodeverk = hentKodeverkFraCacheEllerConsumer(LANDKODER_CACHE_KEY)
        val land = finnFoersteTermForKodeverdi(landkoderKodeverk, landkode)
        return formaterLand(land)
    }

    private fun hentKodeverkFraCacheEllerConsumer(key: String): KodeverkDto? {
        if (skalBrukeCache()) {
            val kodeverk = hentFraCache(key)
            if (kodeverk != null) {
                return kodeverk
            }
        }
        when (key) {
            KOMMUNER_CACHE_KEY -> return kodeverkFraClientEllerCacheSomFallback(key, kodeverkClient.hentKommuner())
            POSTNUMMER_CACHE_KEY -> return kodeverkFraClientEllerCacheSomFallback(key, kodeverkClient.hentPostnummer())
            LANDKODER_CACHE_KEY -> return kodeverkFraClientEllerCacheSomFallback(key, kodeverkClient.hentLandkoder())
        }
        return null
    }

    private fun skalBrukeCache(): Boolean {
        return redisService.getString(KODEVERK_LAST_POLL_TIME_KEY)
            ?.let {
                val lastPollTime = LocalDateTime.parse(it, DateTimeFormatter.ISO_LOCAL_DATE_TIME)
                lastPollTime.plusMinutes(MINUTES_TO_PASS_BETWEEN_POLL).isAfter(LocalDateTime.now())
            } ?: false
    }

    private fun hentFraCache(key: String): KodeverkDto? {
        return redisService.get(key, KodeverkDto::class.java) as? KodeverkDto
    }

    private fun kodeverkFraClientEllerCacheSomFallback(key: String, kodeverk: KodeverkDto?): KodeverkDto? {
        if (kodeverk != null) {
            return kodeverk
        }
        val cachedKodeverk = hentFraCache(key)
        if (cachedKodeverk != null) {
            log.info("Kodeverk-client feilet, men bruker cached kodeverk.")
            return cachedKodeverk
        }
        log.warn("Kodeverk feiler og cache [$key] er tom")
        return null
    }

    private fun finnKodeverdiForFoersteTerm(kodeverk: KodeverkDto?, term: String): String? {
        if (kodeverk == null) {
            return null
        }

        return kodeverk.betydninger?.entries
            ?.firstOrNull {
                val dtos = it.value
                if (dtos.isNotEmpty() && dtos[0].beskrivelser?.get(SPRAAKKODE_NB)?.term != null) {
                    term.equals(dtos[0].beskrivelser?.get(SPRAAKKODE_NB)?.term, ignoreCase = true)
                } else false
            }
            ?.key
    }

    private fun finnFoersteTermForKodeverdi(kodeverk: KodeverkDto?, kodeverdi: String): String? {
        return if (kodeverk != null && kodeverk.betydninger?.containsKey(kodeverdi) == true) {
            kodeverk.betydninger[kodeverdi]?.get(0)?.beskrivelser?.get(SPRAAKKODE_NB)?.term
        } else null
    }

    private fun formaterLand(land: String?): String? {
        if (land == null) {
            return land
        }
        val formaterMedSpace = setUpperCaseBeforeRegex(land.lowercase(Locale.getDefault()), " ")
        val formaterMedDash = setUpperCaseBeforeRegex(formaterMedSpace, "-")
        return setUpperCaseBeforeRegex(formaterMedDash, "/")
    }

    private fun setUpperCaseBeforeRegex(s: String, regex: String): String {
        val split = s.split(regex.toRegex()).toTypedArray()
        val sb = StringBuilder()
        for (i in split.indices) {
            if (i > 0) {
                sb.append(regex)
            }
            if (split[i] == "og") {
                sb.append(split[i])
            } else {
                sb.append(split[i].substring(0, 1).uppercase(Locale.getDefault()))
                sb.append(split[i].substring(1))
            }
        }
        return sb.toString()
    }

    companion object {
        private val log = LoggerFactory.getLogger(KodeverkService::class.java)
        private const val MINUTES_TO_PASS_BETWEEN_POLL: Long = 60

        const val SPRAAKKODE_NB = "nb"
    }
}
